django/core/cache/backends/locmem.py

```diff
@@ -13,117 +13,97 @@
 _locks = {}
 
 
-@contextmanager
-def dummy():
-    """A context manager that does nothing special."""
-    yield
-
-
 class LocMemCache(BaseCache):
     def __init__(self, name, params):
         super().__init__(params)
-        self._cache = _caches.setdefault(name, {})
+        self._cache = _caches.setdefault(name, OrderedDict())
         self._expire_info = _expire_info.setdefault(name, {})
-        self._lock = _locks.setdefault(name, RWLock())
+        self._lock = _locks.setdefault(name, RLock())
 
     def add(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
         key = self.make_key(key, version=version)
         self.validate_key(key)
         pickled = pickle.dumps(value, pickle.HIGHEST_PROTOCOL)
-        with self._lock.writer():
+        with self._lock:
             if self._has_expired(key):
                 self._set(key, pickled, timeout)
                 return True
             return False
 
-    def get(self, key, default=None, version=None, acquire_lock=True):
+    def get(self, key, default=None, version=None):
         key = self.make_key(key, version=version)
         self.validate_key(key)
-        pickled = None
-        with (self._lock.reader() if acquire_lock else dummy()):
+        with self._lock:
             if not self._has_expired(key):
                 pickled = self._cache[key]
-        if pickled is not None:
-            try:
-                return pickle.loads(pickled)
-            except pickle.PickleError:
+                self._cache.move_to_end(key, last=False)
+            else:
+                self._discard(key)
```

Comment:
I don't see a failure if this line is commented out. Sorry if I gave the impression that my previous review was a thorough audit of the test coverage. Can you check more thoroughly?

Reply:
I take this back and have a deeper look at the tests.

Reply:
Which line exactly? If I comment out line 39 with the "move_to_end" call then the test fails.

Reply:
Line 41.  By the way, it would make the patch a bit easier to review and see what has changed if you didn't reorder methods (e.g. set is moved above _set) and delete above _discard). Perhaps the reordering could be done in a separate commit afterwards if it's needed.

Reply:
I don't think that line needs a test. It's not there for correctness, just an optimization. It's there to remove keys if we notice they've expired when we do a "get". It's included simply because the current implementation includes it. Were it not there, expired keys would still move eventually to the end and be removed from the cache.